pkg/controller/worker/state_reconciler.go

```diff
@@ -0,0 +1,140 @@
+	worker := &extensionsv1alpha1.Worker{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, worker); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Deletion flow
+	if worker.DeletionTimestamp != nil {
+		//Nothing to do
+		return reconcile.Result{}, nil
+	}
+
+	// Reconcile flow
+	if err := controller.EnsureFinalizer(r.ctx, r.client, FinalizerName, worker); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	operationType := gardencorev1beta1helper.ComputeOperationType(worker.ObjectMeta, worker.Status.LastOperation)
+	if operationType == gardencorev1beta1.LastOperationTypeCreate {
+		//Something which is not created does not need state update for the moment so we leave it for later
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	if err := r.updateStatusProcessing(r.ctx, worker, operationType, "Updateting the worker state"); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.actuator.Reconcile(r.ctx, worker); err != nil {
+		msg := "Error updateting worker state"
+		utilruntime.HandleError(r.updateStatusError(r.ctx, extensionscontroller.ReconcileErrCauseOrErr(err), worker, operationType, msg))
+		r.logger.Error(err, msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+		return extensionscontroller.ReconcileErr(err)
+	}
+
+	msg := "Successfully update worker state"
+	r.logger.Info(msg, "worker", fmt.Sprintf("%s/%s", worker.Namespace, worker.Name))
+	if err := r.updateStatusSuccess(r.ctx, worker, operationType, msg); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *stateReconciler) updateStatusProcessing(ctx context.Context, worker *extensionsv1alpha1.Worker, lastOperationType gardencorev1beta1.LastOperationType, description string) error {
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, r.client, worker, func() error {
+		worker.Status.LastOperation = extensionscontroller.LastOperation(lastOperationType, gardencorev1beta1.LastOperationStateProcessing, 1, description)
+		return nil
+	})
+}
+
+func (r *stateReconciler) updateStatusError(ctx context.Context, err error, worker *extensionsv1alpha1.Worker, lastOperationType gardencorev1beta1.LastOperationType, description string) error {
+	return extensionscontroller.TryUpdateStatus(ctx, retry.DefaultBackoff, r.client, worker, func() error {
+		worker.Status.ObservedGeneration = worker.Generation
+		worker.Status.LastOperation, worker.Status.LastError = extensionscontroller.ReconcileError(lastOperationType, gardencorev1beta1helper.FormatLastErrDescription(fmt.Errorf("%s: %v", description, err)), 50, gardencorev1beta1helper.ExtractErrorCodes(err)...)
+		return nil
+	})
+}
+
+func (r *stateReconciler) updateStatusSuccess(ctx context.Context, worker *extensionsv1alpha1.Worker, lastOperationType gardencorev1beta1.LastOperationType, description string) error {
```

Comment:
See above, I think we can remove all these update functions to prevent conflicts between the real and the state updater.

Reply:
agree